File: app/src/main/java/com/jiujiu/mvvmTemplate/data/DataManager.java
package com.jiujiu.mvvmTemplate.data;

import com.jiujiu.mvvmTemplate.data.model.Product;
import com.jiujiu.mvvmTemplate.data.repository.ProductRepository;
import com.jiujiu.mvvmTemplate.util.AppConstant;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Maybe;
import io.reactivex.Observable;

@Singleton
public class DataManager {

    private AppPreference mPreference;
    private ProductRepository mProductRepository;

    @Inject
    public DataManager(AppPreference preference, ProductRepository productRepository) {
        mPreference = preference;
        this.mProductRepository = productRepository;
        setCurrentUserName(AppConstant.USERNAME);
    }

    public String getCurrentUserName() {
        return this.mPreference.getUserName();
    }

    public void setCurrentUserName(String userName) {
        this.mPreference.setUserName(userName);
    }

    public void prePupulateData() {
        List<Product> products = new ArrayList<>();
        for (int i = 1; i <= 5; i++) {
            Product p = new Product();
            p.name = "Product Name " + i;
            p.price = i * 10;
            p.brand = "Brand " + i;
            products.add(p);
        }
        mProductRepository.insert(products.toArray(new Product[products.size()]));
    }

    public Observable<List<Product>> getAllProducts() {
        return this.mProductRepository.loadAllProducts();
    }

    public Maybe<Product> getProductById(Long id) {
        return this.mProductRepository.loadProductById(id);
    }
}

File: app/src/main/java/com/jiujiu/mvvmTemplate/data/DataManager.kt
package com.jiujiu.mvvmTemplate.data

import androidx.lifecycle.LiveData
import com.jiujiu.mvvmTemplate.data.model.Product
import com.jiujiu.mvvmTemplate.data.repository.ProductRepository
import com.jiujiu.mvvmTemplate.util.AppConstant
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DataManager @Inject constructor(
        private val mPreference: AppPreference,
        private val mProductRepository: ProductRepository
) {

    var currentUserName: String?
        get() = this.mPreference.userName
        set(userName) {
            this.mPreference.userName = userName
        }

    val allProducts: LiveData<List<Product>>
        get() = this.mProductRepository.loadAllProducts()

    init {
        currentUserName = AppConstant.USERNAME
    }

    suspend fun prePopulateData() {
        withContext(Dispatchers.IO){
            val products = arrayListOf<Product>().apply {
                for (i in 1..5) {
                    add(Product(null, "product Name $i", "Brand $i", i * 10.0))
                }
            }
            mProductRepository.insert(*products.toTypedArray())
        }
    }

    fun getProductById(id: Long?): LiveData<Product> {
        return this.mProductRepository.loadProductById(id)
    }
}
